### PR TITLE
feat/jinja templates

### PR DESCRIPTION
**Pull request by mandresm**
After this PR, files that are handled by ESM-Tools can be `jinja2` templates and will be rendered at the time of the copying.

This was a request by @JanStreffing to be able to handle variable substitution in `xml` files. Jan, could you test it? You can do your xml and yaml modifications in a new branch departing from this one. We can merge in release that branch right after this one gets merged, and there should be no conflicts.

### How does it work

ESM-Tools automatically tries to resolve jinja templates during movements. It searches for sources finishing with `.j2` and then it knows that instead of copying/linking/moving it needs to render the template and produce the new target file. We pass `config` to `jinja`, that means you just need to specify the variable you want to subsitute in the right format with the full path separated by points:

**xios.yaml**
```diff
config_sources:
-        domain_def: ${xml_dir}/domain_def.xml
+        domain_def: ${xml_dir}/domain_def.j2
        field_def: ${xml_dir}/field_def.xml
        file_def: ${xml_dir}/file_def.xml
```
**domain_def.xml**
```xml
<domain_definition>
  <!-- Definition of the native domain of the model -->
  <domain id="reduced_gaussian" long_name="reduced Gaussian grid" type="gaussian" />

  <!-- Definition of regular Gaussian domains -->
  <domain_group id="regular_domains" type="rectilinear" >
    <domain id="regular" long_name="regular grid" ni_glo="{{xios.ni_glo}}" nj_glo="{{xios.nj_glo}}" >
      <generate_rectilinear_domain />
      <interpolate_domain order="1" write_weight="true" />
    </domain>
  </domain_group>
</domain_definition>
```

or with python dict syntax

```xml
<domain_definition>
  <!-- Definition of the native domain of the model -->
  <domain id="reduced_gaussian" long_name="reduced Gaussian grid" type="gaussian" />

  <!-- Definition of regular Gaussian domains -->
  <domain_group id="regular_domains" type="rectilinear" >
    <domain id="regular" long_name="regular grid" ni_glo="{{xios['ni_glo']}}" nj_glo="{{xios['nj_glo']}}" >
      <generate_rectilinear_domain />
      <interpolate_domain order="1" write_weight="true" />
    </domain>
  </domain_group>
</domain_definition>
```

**TODO**

- [x] Add documentation
- [x] Test
- [x] Error handling for not defined extension

### CODE REVIEW

**Comment by mandresm** [comment]
Document that tarjet changes extension

File: src/esm_runscripts/filelists.py [hunk 1]
```diff
@@ -1,3 +1,4 @@
+import concurrent
```
Comment (pgierz) [inline src/esm_runscripts/filelists.py]: Not to be forgotten: The parallel movements need to be merged first.

File: src/esm_runscripts/filelists.py [hunk 2]
```diff
@@ -999,59 +1012,72 @@ def copy_files(config, filetypes, source, target):
...
+                        if os.path.isfile(file_target) and filecmp.cmp(
+                            file_source, file_target
+                        ):
+                            logger.debug(
+                                f"Source and target file are identical, skipping {file_source}",
+                            )
+                            helpers.print_datetime(config)
+                            continue
+                        files_to_be_moved.append(
+                            {
```
Comment (pgierz) [inline src/esm_runscripts/filelists.py]: I don't understand this change. I assume it is needed for parallel copying to work correctly?

File: src/esm_runscripts/jinja.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+            "ni_glo": 400,
+            "nj_glo": 600,
+        },
+    }
+
+    render_template(
+        config,
+        "/Users/mandresm/Codes/esm_tools/namelists/oifs/43r3/xios/domain_def.j2",
+        "out_file.xml",
+    )
```
Comment (pgierz) [inline src/esm_runscripts/jinja.py]: You could consider moving this into a unit test (if you get bored and need 15 minutes to do something for esm-tools)

**Review by pgierz — APPROVED** [review]
Looks good from the Python side, still needs user-testing (@JanStreffing) and documentation in the handbook. Nice work! 👍🏻

**Comment by JanStreffing** [comment]
> Looks good from the Python side, still needs user-testing (@JanStreffing) and documentation in the handbook. Nice work! 👍🏻

I hope to give this a try tomorrow. Maybe I just drop by one of your offices?

**Comment by pgierz** [comment]
Sure, I'll be around tomorrow

File: src/esm_runscripts/filelists.py (hunk 2 above)
Comment (mandresm) [reply src/esm_runscripts/filelists.py]: alreaday discussed in person

**Comment by mandresm** [comment]
Closes #511

**Comment by mandresm** [comment]
#bump

**Comment by mandresm** [comment]
@JanStreffing, you can read about how to use this feature here: https://esm-tools.readthedocs.io/en/v6.41.0/jinja.html

**Pull request merged**